indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/IndyUser.kt

```diff
@@ -1,596 +1,201 @@
 package com.luxoft.blockchainlab.hyperledger.indy
 
-import com.luxoft.blockchainlab.hyperledger.indy.helpers.TailsHelper
+import com.luxoft.blockchainlab.hyperledger.indy.ledger.LedgerUser
 import com.luxoft.blockchainlab.hyperledger.indy.models.*
-import com.luxoft.blockchainlab.hyperledger.indy.roles.IndyIssuer
-import com.luxoft.blockchainlab.hyperledger.indy.roles.IndyProver
-import com.luxoft.blockchainlab.hyperledger.indy.roles.IndyTrustee
-import com.luxoft.blockchainlab.hyperledger.indy.roles.IndyVerifier
 import com.luxoft.blockchainlab.hyperledger.indy.utils.SerializationUtils
-import com.luxoft.blockchainlab.hyperledger.indy.utils.getRootCause
-import mu.KotlinLogging
-import org.hyperledger.indy.sdk.anoncreds.Anoncreds
-import org.hyperledger.indy.sdk.anoncreds.DuplicateMasterSecretNameException
-import org.hyperledger.indy.sdk.did.Did
-import org.hyperledger.indy.sdk.pairwise.Pairwise
-import org.hyperledger.indy.sdk.pool.Pool
-import org.hyperledger.indy.sdk.wallet.Wallet
-import org.hyperledger.indy.sdk.wallet.WalletItemNotFoundException
-import java.util.concurrent.ExecutionException
+import com.luxoft.blockchainlab.hyperledger.indy.wallet.WalletUser
 
 
 /**
  * The central class that encapsulates Indy SDK calls and keeps the corresponding state.
- *
- * Create one instance per each server node that deals with Indy Ledger.
+ * This is implementation of [SsiUser] so it should support every valid [LedgerUser] and [WalletUser]
+ *  implementation.
  */
-open class IndyUser(
-    val pool: Pool,
-    val wallet: Wallet,
-    did: String?,
-    didConfig: String = EMPTY_OBJECT,
-    val tailsPath: String = "tails",
-    val masterSecretId: String = "main"
-) : IndyIssuer, IndyProver, IndyTrustee {
-
-    private val logger = KotlinLogging.logger {}
-
-    override val did: String
-    val verkey: String
-    val ledgerService: LedgerService
+class IndyUser(
+    override val walletUser: WalletUser,
+    override val ledgerUser: LedgerUser
+) : SsiUser {
```

Comment:
Why dont we have constructor from Wallet?
val walletUser = IndySDKWalletUser(wallet)
val ledgerUser = IndyPoolLedgerUser(pool, walletUser.did) { walletUser.sign(it) }

Reply:
Because `IndyUser` doesn't know about `IndySDKWalletUser` or `IndyPoolLedgerUser`. It knows only about interfaces which don't possess any constructor at all.